pupil-spa/src/app/services/connection-test/connection-test.service.ts

```diff
@@ -0,0 +1,174 @@
+        this.fibonacci(this.fibonacciN);
+      }
+
+      this.processingTime = Date.now() - startTime;
+      resolve();
+    });
+  }
+
+  private getFileUrl(fileSize: string): string {
+    // strip the trailing '/' if there is any
+    const connectionTestBlobUrl = APP_CONFIG.connectionTestBlobUrl.replace(/\/$/, '');
+
+    return `${connectionTestBlobUrl}/${APP_CONFIG.connectionTestBlobStorageName}/${fileSize}.text?nocache=${(new Date()).getTime()}`;
+  }
+
+  private requestFile(url: string): Promise<any> {
+    const startTime = Date.now();
+
+    return this.http
+      .get(url, { responseType: 'text', observe: 'response' })
+      .pipe(
+        map((resp: any) => {
+          const fileSize = resp.headers.keys().includes('content-length')
+            ? resp.headers.get('content-length')
+            : resp.body.length;
+          const downloadTime = Date.now() - startTime;
+
+          return { downloadTime, fileSize };
+        })
+      )
+      .toPromise();
+  }
+
+  public benchmarkConnection(): Promise<any> {
+    return new Promise(async (resolve) => {
+      let downloadTime, currentFileSize, multiplier;
+
+      for (let i = 0; i < this.filesSizes.length; i++) {
+        // dealing with kb or mb
+        multiplier = this.filesSizes[i].endsWith('kb') ? 1024 : 1024 * 1024;
+        const fileUrl = this.getFileUrl(this.filesSizes[i]);
+
+        let data;
+        try {
+          data = await this.requestFile(fileUrl);
+        } catch (e) {
+          continue;
+        }
+
+        currentFileSize = parseInt(this.filesSizes[i].replace(/\D/g, ''), 10);
+        downloadTime = data.downloadTime;
+
+        // retry downloading the file if the fileSize doesn't correspond
+        // to what we expected here
+        if (downloadTime < this.downloadTimeTimeout && data.fileSize !== multiplier * currentFileSize) {
+          i--;
+        }
+        if (downloadTime > this.downloadTimeTimeout) {
+          break;
+        }
+      }
+
+      this.connectionSpeed = currentFileSize * multiplier / downloadTime;
+      resolve();
+    });
+  }
+
+  private submitTest(testResults: object): Promise<void> {
+    return new Promise((resolve, reject) => {
+      if (this.queueName === '') {
+        return reject();
+      }
+
+      const queueService = this.getQueueService();
+      const encoder = new AzureStorage.Queue.QueueMessageEncoder.TextBase64QueueMessageEncoder();
+
+      const message = JSON.stringify(testResults);
+      const encodedMessage = encoder.encode(message);
+      queueService.createMessage(this.queueName, encodedMessage, function (error, result, response) {
+        if (error) {
+          return reject();
+        }
+
+        resolve();
+      });
+    });
+  }
+
+  private getQueueService() {
+    return AzureStorage.Queue.createQueueServiceWithSas(
+      APP_CONFIG.testSasUrl,
+      APP_CONFIG.testSasToken
+    ).withFilter(new AzureStorage.Queue.ExponentialRetryPolicyFilter());
```

Comment:
@danwilliams is there a reason you chose exponential instead of linear policy filter?

Reply:
Mmmm, it just seemed best to me... I guess it's somewhat arbitrary?

Reply:
@danwilliams @GeorgeChatzigiannis https://azure.github.io/azure-storage-node/common_filters_exponentialretrypolicyfilter.js.html